provider/hetzner.go

```diff
@@ -0,0 +1,206 @@
+/*
+	by Vladimir Smagin, 2020
+	vlad@blindage.org
+*/
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	hclouddns "git.blindage.org/21h/hcloud-dns"
+	log "github.com/sirupsen/logrus"
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+)
+
+const (
+	hetznerCreate = "CREATE"
+	hetznerDelete = "DELETE"
+	hetznerUpdate = "UPDATE"
+	hetznerTTL    = 600
+)
+
+type HetznerChanges struct {
+	Action            string
+	ZoneID            string
+	ZoneName          string
+	ResourceRecordSet hclouddns.HCloudRecord
+}
+
+type HetznerProvider struct {
+	Client       hclouddns.HCloudDNS
+	domainFilter endpoint.DomainFilter
+	DryRun       bool
+}
+
+func NewHetznerProvider(ctx context.Context, domainFilter endpoint.DomainFilter, dryRun bool) (*HetznerProvider, error) {
+	token, ok := os.LookupEnv("HETZNER_TOKEN")
+	if !ok {
+		return nil, fmt.Errorf("No token found")
+	}
+
+	provider := &HetznerProvider{
+		Client:       *hclouddns.New(token),
+		domainFilter: domainFilter,
+		DryRun:       dryRun,
+	}
+	return provider, nil
+}
+
+func (p *HetznerProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	zones, err := p.Client.GetZones(hclouddns.HCloudGetZonesParams{})
+	if err != nil {
+		return nil, err
+	}
+	endpoints := []*endpoint.Endpoint{}
+	for _, zone := range zones.Zones {
+		records, err := p.Client.GetRecords(hclouddns.HCloudGetRecordsParams{ZoneID: zone.ID})
+		if err != nil {
+			return nil, err
+		}
+
+		for _, r := range records.Records {
+			if supportedRecordType(string(r.RecordType)) {
+				name := r.Name + "." + zone.Name
+
+				if r.Name == "@" {
+					name = zone.Name
+				}
+
+				endpoints = append(endpoints, endpoint.NewEndpoint(name, string(r.RecordType), r.Value))
+			}
+		}
+	}
+
+	return endpoints, nil
+}
+
+func (p *HetznerProvider) ApplyChanges(ctx context.Context, changes *plan.Changes) error {
+	combinedChanges := make([]*HetznerChanges, 0, len(changes.Create)+len(changes.UpdateNew)+len(changes.Delete))
+
+	combinedChanges = append(combinedChanges, p.newHetznerChanges(hetznerCreate, changes.Create)...)
+	combinedChanges = append(combinedChanges, p.newHetznerChanges(hetznerUpdate, changes.UpdateNew)...)
+	combinedChanges = append(combinedChanges, p.newHetznerChanges(hetznerDelete, changes.Delete)...)
+
+	return p.submitChanges(ctx, combinedChanges)
+}
+
+func (p *HetznerProvider) submitChanges(ctx context.Context, changes []*HetznerChanges) error {
+	if len(changes) == 0 {
+		log.Infof("All records are already up to date")
```

Comment:
What do you think to use Debug log level instead ?

Reply:
Nevermind. it seems this is the best way to mark as info, base on comment https://github.com/kubernetes-sigs/external-dns/pull/1573#issuecomment-627298237